### PR TITLE
Add climbing rope training to athletics

### PR DESCRIPTION
**Pull request by jandersson**
This implements support for the climbing ropes that allow for stationary training of athletics. The script makes the assumption that you have zills.

If you set `have_climbing_rope` to true, then the script will go to your safe room, play a song on your zills, and begin climb practice on the rope. Playing a song is necessary to make the rope climbable, I guess its like a snake-charmer type of deal.

The script attempts to maximize performance training while still training athletics. It starts by determining the best performance training song (either by copying `UserVars.song` or starting from scratch). While climbing, if the song is too hard/easy, the song difficulty is stepped down/up.

Item details and usage are documented here:
* item: https://elanthipedia.play.net/Item:Dark_climbing_rope_woven_with_strangely_glowing_threads
* usage: https://elanthipedia.play.net/Dancing_rope

### CODE REVIEW

File: profiles/base.yaml [hunk 1]
```diff
@@ -564,16 +564,4 @@ predict_event: false
 
 learned_column_count: 2
 sort_auto_head: true
-
-scroll_stackers:
```
Comment (rpherbig) [inline profiles/base.yaml]: Why did you remove these settings?

File: athletics.lic [hunk 2]
```diff
@@ -2,22 +2,26 @@
   Documentation: https://elanthipedia.play.net/Lich_script_repository#athletics
```
Comment (rpherbig) [inline athletics.lic]: This script isn't integrated with `crossing-training`. Do you plan to also update the athletics logic in `crossing-training` to support the climbing rope?

File: athletics.lic [hunk 3]
```diff
@@ -121,19 +126,21 @@ class Athletics
         end
         return if remaining_attempts.zero?
       end
-      Flags.add('ct-climbing-finished', 'You finish practicing your climbing')
+      Flags.add('ct-climbing-finished', 'You finish practicing your climbing', 'The rope\'s will quickly fades away')
       Flags.add('ct-climbing-combat', 'You are engaged in combat')
-      bput("climb practice #{target}", 'You begin to practice ')
+      fix_standing
+      bput("play #{UserVars.song} on zills", 'You begin', 'You\'re already playing a song') if target.eql?('climbing rope')
```
Comment (rpherbig) [inline athletics.lic]: There's a ton of logic in `crossing-training::play_song?` that we're missing here (i.e. edge cases that are handled). Should that be extracted and reused here?

**Comment by BardWyrrd** [comment]
Script fails if Performance is enough higher than Athletics that User.Vars.song used to train Performance makes the rope too hard to climb with current Athletics skill. When this happens, script stays active after the failed climb message. See below. Is there any way to flex the song used to match Athletics?

    Performance:    421 93.71% thoughtful     (4/34)
       Athletics:    261 39.69% understanding (14/34)

--- Lich: go2 has exited.
[athletics]>get climbing rope
>
You get a dark climbing rope woven with strangely glowing threads from inside your tapestry carpetbag.
>
[athletics]>play tarantella on zills
>
You begin a spritely tarantella on your copper zills.
>
[athletics]>climb practice climbing rope
Directing your attention toward your rope, the ends immediate begin twitching.  The rope suddenly unfurls itself completly, moving in pace with the music.  You hang on tightly as a dark climbing rope woven with strangely glowing threads finishes its transformation, and reaching up with your free hand you begin to climb.
>
This climb is too difficult, so you stop practicing.  Your climbing rope coils back up neatly in your hand.

File: profiles/base.yaml (hunk 1 above)
Comment (jandersson) [inline profiles/base.yaml]: Yikes

File: athletics.lic (hunk 2 above)
Comment (jandersson) [inline athletics.lic]: Yea I would like to see crossing-trainer use the rope. It might make sense to get the logic in athletics working and then have crossing-training call the athletics script if the rope is being used.

File: athletics.lic (hunk 3 above)
Comment (jandersson) [inline athletics.lic]: Ill take a look at copying or commonizing it

**Comment by jandersson** [comment]
Will have to determine the song outside of just straight up using the `song` uservar then

File: athletics.lic [hunk 4]
```diff
@@ -2,22 +2,26 @@
...
 
     @climbing_target = get_data('athletics').practice_options[@settings.climbing_target]
+    have_climbing_rope = get_settings.have_climbing_rope
 
     start_exp = DRSkill.getxp('Athletics')
     @end_exp = [start_exp + 15, 29].min
 
-    if @climbing_target
+    if have_climbing_rope
+      climb_practice(get_settings.safe_room, 'climbing rope', false)
```
Comment (jandersson) [inline athletics.lic]: I'm realizing that it might be easier to have the rope setup/usage/cleanup all in its own methods rather than trying to piggy-back in `climb_practice`

**Comment by jandersson** [comment]
I think what id like to try is playing your performance training song from `UserVars.song` and then if the climb fails, start stepping the song difficulty back until it works.

**Comment by jandersson** [comment]
In the meantime I'll add in a user setting to let the user specify what song to play... if its undefined then it uses your performance training song

File: athletics.lic (hunk 3 above)
Comment (jandersson) [inline athletics.lic]: I'm copying it and cutting out crossing-trainer specific lines. It seems like this logic can be commonized, but I'll save that for a future commit or PR.

**Comment by rpherbig** [comment]
Crossing-training has some logic to zero in on the right song to play. You could copy that over and just change the success/fail conditions.

**Comment by jandersson** [comment]
My plan is to copy it over and re-use the logic to determine the best performance training song and store this as the initial value in `UserVars.athletics_song`. If the climb is too difficult, then step down the song style to a lower difficulty and retry.. if too easy step it up.

File: athletics.lic [hunk 5]
```diff
@@ -35,6 +39,111 @@ class Athletics
...
+      return play_climbing_song?
+    when 'You struggle to begin'
+      if UserVars.song != @song_list.first.first
+        stop_play
+        UserVars.song = @song_list.first.first
+        return play_climbing_song?
+      end
+    end
+
+    return true unless blocking
```
Comment (jandersson) [inline athletics.lic]: Remove the blocking check and parameter, not interested in blocking here.

File: athletics.lic [hunk 6]
```diff
@@ -35,6 +39,111 @@ class Athletics
...
+      pause 1
+      waitrt?
+    end
+
+    bput('wear my zills', 'You slide')
+    bput("stow my #{cloth}", 'You put')
+  end
+
+  def play_climbing_song?(blocking = false)
+    UserVars.song = @song_list.first.first unless UserVars.song
```
Comment (jandersson) [inline athletics.lic]: change to `climbing_song` to differentiate from performance training song

File: athletics.lic [hunk 7]
```diff
@@ -35,6 +39,111 @@ class Athletics
...
+
+  def train_with_rope
+    return unless exists?('climbing rope')
+    return unless exists?('zills')
+
+    # TODO: Move setup logic to its own method
+    Flags.add('climbing-finished', 'You finish practicing your climbing', 'The rope\'s will quickly fades away')
+    stow_hands
+    @song_list = get_data('perform').perform_options
+    climbing_song = UserVars.song
```
Comment (jandersson) [inline athletics.lic]: Unused, remove.

File: athletics.lic [hunk 8]
```diff
@@ -35,6 +39,106 @@ class Athletics
   def done_training?
     DRSkill.getxp('Athletics') >= @end_exp
   end
+  
+  def stop_play
+    bput('stop play', 'You stop playing your song', 'In the name of', "But you're not performing")
+    Flags['ct-song'] = true
```
Comment (jandersson) [inline athletics.lic]: Is there any reason to keep this?

Comment (rpherbig) [inline athletics.lic]: Not right now, but we should revisit this if `athletics` is integrated into `crossing-training`.

File: athletics.lic [hunk 9]
```diff
@@ -121,14 +222,13 @@ class Athletics
         end
         return if remaining_attempts.zero?
       end
-      Flags.add('ct-climbing-finished', 'You finish practicing your climbing')
+      Flags.add('ct-climbing-finished', 'You finish practicing your climbing', 'The rope\'s will quickly fades away')
```
Comment (rpherbig) [inline athletics.lic]: Should these flags have the `ct` prefix? That usually indicates they are used in `crossing-training`.

File: athletics.lic [hunk 10]
```diff
@@ -121,14 +222,13 @@ class Athletics
...
       Flags.add('ct-climbing-combat', 'You are engaged in combat')
       bput("climb practice #{target}", 'You begin to practice ')
       loop do
         pause
-        break if Flags['ct-climbing-finished']
         return if Flags['ct-climbing-combat']
-        break if done_training?
+        break if done_training? || Flags['ct-climbing-finished']
```
Comment (rpherbig) [inline athletics.lic]: I kind of liked these being split out, but it's not a big deal

**Comment by jandersson** [comment]
@BardWyrrd The latest update attempts to find the right song to play based on messaging, give it a try when you can. It worked on my bard who has 600 performance 300 athletics.

**Comment by jandersson** [comment]
```
This climb is too difficult, so you stop practicing.  Your climbing rope coils back up neatly in your hand.                                                                                                                            
[athletics]>stop play                                                                                                                                                                                                                  
You stop playing your song.                                                                                                                                                                                                            
[athletics]>play minuet halt                                                                                                                                                                                                           
You begin a halting minuet on your copper zills.                                                                                                                                                                                       
[athletics]>climb practice climbing rope                                                                                                                                                                                               
Directing your attention toward your rope, the ends immediate begin twitching.  The rope suddenly unfurls itself completly, moving in pace with the music.  You hang on tightly as a dark climbing rope woven with strangely glowing   
  threads finishes its transformation, and reaching up with your free hand you begin to climb.                                                                                                                                         
>                                                                                                                                                                                                                                      
A light pleasant breeze continues to swirl around you.                                                                                                                                                                                 
>                                                                                                                                                                                                                                      
Mimz pulls and tugs on her grass, testing it thoroughly.                                                                                                                                                                               
Mimz proudly examines her new braided heavy rope.                                                                                                                                                                                      
>                                                                                                                                                                                                                                      
Mimz drops a braided heavy rope.                                                                                                                                                                                                       
>                                                                                                                                                                                                                                      
[Mimz is summoning the janitor to this room.  It will arrive shortly.  To stop the janitor type DUMP REJECT now.]                                                                                                                      
>                                                                                                                                                                                                                                      
Mimz gets some quilted gloves from inside her duffel bag.                                                                                                                                                                              
>                                                                                                                                                                                                                                      
Mimz slips some quilted gloves onto her hands.                                                                                                                                                                                         
>                                                                                                                                                                                                                                      
You continue playing on your copper zills.                                                                                                                                                                                             
>                                                                                                                                                                                                                                      
This climb is too difficult, so you stop practicing.  Your climbing rope coils back up neatly in your hand.                                                                                                                            
[athletics]>stop play                                                                                                                                                                                                                  
You stop playing your song.                                                                                                                                                                                                            
[athletics]>play serenade                                                                                                                                                                                                              
You begin a quiet serenade on your copper zills.                                                                                                                                                                                       
[athletics]>climb practice climbing rope                                                                                                                                                                                               
Directing your attention toward your rope, the ends immediate begin twitching.  The rope suddenly unfurls itself completly, moving in pace with the music.  You hang on tightly as a dark climbing rope woven with strangely glowing   
  threads finishes its transformation, and reaching up with your free hand you begin to climb.                                                                                                                                         
>                                                                                                                                                                                                                                      
```

File: athletics.lic (hunk 8 above)
Comment (jandersson) [inline athletics.lic]: I'll remove it for now

**Comment by jandersson** [comment]
This PR is in a working state for 3 characters that I've tried it on.
